### ISSUE TITLE
Suport Vector Sum Aggregation

### ISSUE THREAD

**Issue by ajdaling** [pgvector/pgvector#141]
First, love this extension and I'm very happy RDS started supporting it.

Apologies if this exists and I missed it but it would be great to add support for an aggregate sum function. There are some use cases where it is preferable to sum embedding vectors. Additionally, it would be very convenient to be able to use this extension for general purpose vector operations (as opposed to custom array_agg functions/aggregates).

**Comment by ankane** [comment]
Hi @ajdaling, thanks for the suggestion. Can you give some specific examples of use cases?

**Comment by ankane** [comment]
Since neither seems common, it's not something I'd like to add right now, but happy to revisit if other use cases come up.

**Comment by ajdaling** [comment]
Totally understand. Thanks for the quick response.

**Comment by mikeizbicki** [comment]
PR #13 from 2 years ago has working implementations of these routines along with test cases, but was not merged. @ajdaling the code in that pr should be a drop in replacement for this repo that does what you want.

@ankane I realize there is legitimate cognitive overhead to including functions that you didn't write in the library, but since these don't affect the "core" functionality of the library at all, I think they'd be relatively easy to merge.  There have been a handful of issues now that you've closed because the use cases of this feature are "not common", but perhaps the fact that so many different people have asked for this feature implies that this is actually a commonly sought feature worth the mental overhead on your part.

**Comment by ajdaling** [comment]
In case it is useful, I did test out that PR on a non-RDS instance and it seems to outperform my custom functions so I think there is virtue in using the pgvector implementation in PR #13 

For reference, my simple sum function is 
`CREATE OR REPLACE FUNCTION array_add(_a numeric[], _b numeric[])
  RETURNS numeric[]
AS
$$
BEGIN
  RETURN ARRAY(
    SELECT coalesce(a, 0) + coalesce(b, 0)
    FROM unnest(_a, _b) WITH ORDINALITY AS x(a, b, n)
    ORDER BY n
  );
END
$$ LANGUAGE plpgsql;

CREATE AGGREGATE sum_array_agg(numeric[])
(
  sfunc = array_add,
  stype = numeric[],
  initcond = '{}'
);`

**Comment by jkatz** [comment]
@ajdaling  Here is an example that uses PL/Rust that from my testing, will have generally better performance than the PL/pgSQL equivalent. Please see the experiment below:

```sql
-- function used to generate test data
CREATE OR REPLACE FUNCTION generate_random_vector(dim int)
RETURNS float4[]
LANGUAGE plpgsql
STRICT
AS $$
    DECLARE
        i int := 1;
        v float4[];
    BEGIN
        FOR i IN 1..dim LOOP
            v[i] := random();
        END LOOP;

        RETURN v;
    END;
$$;

-- create the extensions
CREATE EXTENSION IF NOT EXISTS plrust;
CREATE EXTENSION IF NOT EXISTS vector;

-- generate test data
CREATE TABLE vectors (
    id int PRIMARY KEY,
    embedding vector(768)
);

INSERT INTO vectors
SELECT n, generate_random_vector(768)
FROM generate_series (1,1000000) n;

-- prior to PL/Rust 1.2.0, you can use this function to create the accumulator method:
CREATE OR REPLACE FUNCTION public.vector_add(x float4[], y float4[])
RETURNS float4[]
LANGUAGE plrust
IMMUTABLE PARALLEL SAFE STRICT
AS $$
    match x.len() {
        0 => {
            Ok(Some(y.into_iter().collect()))
        }
        _ => {
            let mut iter = y.into_iter().zip(x.into_iter());

            let z: Vec<Option<f32>> = iter.map(|(xi, yi)| Some(xi.unwrap() + yi.unwrap()) ).collect();

            Ok(Some(z))
        }
    }
$$;

-- With PL/Rust 1.2.0 and higher, you can use the following function, which can be ~25% more performant:
CREATE OR REPLACE FUNCTION public.vector_add(x float4[], y float4[])
    RETURNS float4[]
    LANGUAGE plrust
    IMMUTABLE PARALLEL SAFE STRICT
AS $$
    match x.len() {
        0 => {
            Ok(Some(y.into_iter().collect()))
        }
        _ => {
            let z = y.as_slice().into_iter()
                .zip(
                    x.as_slice().into_iter()
                )
                .map(|(xi, yi)| Some(xi + yi))
                .collect();

            Ok(Some(z))
        }
    }
$$;

-- this creates the custom aggreagtor:
CREATE OR REPLACE AGGREGATE vector_agg(float4[])
(
    sfunc = vector_add,
    stype = float4[],
    initcond = '{}'
);

-- test!
EXPLAIN ANALYZE SELECT vector_agg(embedding)
FROM vectors;
```

To compare, I took your example and optimized it (`float4` will be more performant than `numeric`). In my tests, I found the PL/Rust equivalent was 2-3x more performant.

```sql
CREATE OR REPLACE FUNCTION vector_add_plpgsql(_a float4[], _b float4[])
RETURNS float4[]
AS
$$
BEGIN
    IF array_length(_a, 1) = 0 THEN
        RETURN _b;
    END IF;

    RETURN ARRAY(
    SELECT coalesce(a, 0) + coalesce(b, 0)
    FROM unnest(_a, _b) WITH ORDINALITY AS x(a, b, n)
    ORDER BY n
);
END
$$ LANGUAGE plpgsql;

CREATE OR REPLACE AGGREGATE vector_agg_plpgsql(float4[])
(
    sfunc = vector_add_plpgsql,
    stype = float4[],
    initcond = '{}'
);
```

Hope this helps!

**Comment by ankane** [comment]
Just fyi, this was added in ed1bc3e856d1726426217c2e71e8681a47856f55 and will be included in 0.5.0. Thanks again for sharing your use cases @ajdaling.
